### PR TITLE
Rev ref returns

### PR DESCRIPTION
**Pull request by PhrygianGates**
This PR rebase the https://github.com/vgvassilev/clad/pull/425 pull-request on top of master.

### CODE REVIEW

**Comment by parth-07** [comment]
@PhrygianGates Can you please debug and fix the failing debug checks?

**Comment by parth-07** [comment]
Is the pull request ready for review?

**Comment by PhrygianGates** [comment]
> @PhrygianGates Can you please debug and fix the failing debug checks?

I'm not sure as it seems to go wrong when installing something. I'll push again to see what happens.

**Comment by vgvassilev** [comment]
> > @PhrygianGates Can you please debug and fix the failing debug checks?
> 
> I'm not sure as it seems to go wrong when installing something. I'll push again to see what happens.

This fails elsewhere due to some server outage it seems. We can ignore this for now.

File: include/clad/Differentiator/ReverseModeForwPassVisitor.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "clang/Sema/Sema.h"
+
+#include "llvm/ADT/SmallVector.h"
+
+namespace clad {
+class ReverseModeForwPassVisitor : public ReverseModeVisitor {
+private:
+  Stmts m_Globals;
+
+  llvm::SmallVector<clang::QualType, 8>
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/ReverseModeForwPassVisitor.h:17]: We should be able to suppress these by adding `IgnorePowersOf2IntegerValues=true` in the clang-tidy config: https://clang.llvm.org/extra/clang-tidy/checks/readability/magic-numbers.html

File: lib/Differentiator/DerivativeBuilder.cpp [hunk 2]
```diff
@@ -22,6 +22,9 @@
 #include "clad/Differentiator/HessianModeVisitor.h"
 #include "clad/Differentiator/JacobianModeVisitor.h"
 #include "clad/Differentiator/ReverseModeVisitor.h"
+#include "clad/Differentiator/ReverseModeForwPassVisitor.h"
+
```
Comment (parth-07) [inline lib/Differentiator/DerivativeBuilder.cpp:26]: This new-line can be removed.

File: lib/Differentiator/ReverseModeForwPassVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+  for (Stmt* S : ReverseModeVisitor::m_Globals)
+    addToCurrentBlock(S);
+
+  if (auto CS = dyn_cast<CompoundStmt>(forward))
+    for (Stmt* S : CS->body())
+      addToCurrentBlock(S);
+
+  Stmt* fnBody = endBlock();
+  // llvm::errs() << "Derive: dumping fnBody:\n";
```
Comment (parth-07) [inline lib/Differentiator/ReverseModeForwPassVisitor.cpp:75]: Remove `llvm::errs()` comments from here and other places.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 4]
```diff
@@ -2394,8 +2479,9 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
       Expr* assignDerivativeE =
           BuildOp(BinaryOperatorKind::BO_Assign, derivedVDE,
                   BuildOp(UnaryOperatorKind::UO_AddrOf,
-                          initDiff.getForwSweepExpr_dx()));
+                          initDiff.getForwSweepExpr_dx()/*bug here*/ ));
       addToCurrentBlock(assignDerivativeE);
+      // bug here?
```
Comment (parth-07) [inline lib/Differentiator/ReverseModeVisitor.cpp:2484]: Remove this.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 5]
```diff
@@ -1626,6 +1626,30 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
             BuildOp(UnaryOperatorKind::UO_AddrOf, baseDiff.getExpr_dx()));
       }
 
+      /// Add base derivative expression in the derived call output args list if
```
Comment (parth-07) [inline lib/Differentiator/ReverseModeVisitor.cpp:1629]: Can this comment be removed?

**Review by parth-07 — CHANGES_REQUESTED** [review]

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 6]
```diff
@@ -1857,15 +1861,82 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+                                  BuildDeclRef(calleeFnForwPassFD), noLoc,
+                                  CallArgs, noLoc)
+                  .get();
+      }
+      auto callRes = StoreAndRef(call);
+      auto resValue =
+          utils::BuildMemberExpr(m_Sema, getCurrentScope(), callRes, "value");
+      auto resAdjoint =
+          utils::BuildMemberExpr(m_Sema, getCurrentScope(), callRes, "adjoint");
+      return StmtDiff(resValue, resAdjoint, resAdjoint);
```
Comment (parth-07) [inline lib/Differentiator/ReverseModeVisitor.cpp:1927]:
```suggestion
      return StmtDiff(resValue, nullptr, resAdjoint);
```
Call expressions do not have a corresponding derivative that can be used explicitly in the reverse-sweep, 
as we insert the derivative call expression implicitly.

Comment (parth-07) [reply lib/Differentiator/ReverseModeVisitor.cpp:1927]: You will also need to change, `ReverseModeVisitor::DifferentiateVarDecl`:

```cpp
if (isDerivativeOfRefType) {
        initDiff = Visit(VD->getInit());
        if (!initDiff.getExpr_dx()) {
          VDDerivedType =
              ComputeAdjointType(VD->getType().getNonReferenceType());
          isDerivativeOfRefType = false;
        }
        VDDerivedInit = getZeroInit(VDDerivedType);
      }
```
to:

```cpp
if (isDerivativeOfRefType) {
        initDiff = Visit(VD->getInit());
        if (!initDiff.getForwSweepExpr_dx()) { // This line changed!!
          VDDerivedType =
              ComputeAdjointType(VD->getType().getNonReferenceType());
          isDerivativeOfRefType = false;
        }
        VDDerivedInit = getZeroInit(VDDerivedType);
      }
```

**Review by parth-07 — CHANGES_REQUESTED** [review]

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 7]
```diff
@@ -1857,15 +1861,84 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+      // Recreate the original call expression.
+      call = m_Sema
+                      .ActOnCallExpr(getCurrentScope(),
+                                      Clone(CE->getCallee()),
+                                      noLoc,
+                                      CallArgs,
+                                      noLoc)
+                      .get();
+      return StmtDiff(call);
+    }
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/ReverseModeVisitor.cpp:1940]: Can we incorporate that suggestion?

**Review by vgvassilev — COMMENTED** [review]
Most of the clang-tidy suggestions make sense. Could you incorporate them in the PR?

File: demos/ComputerGraphics/smallpt/SmallPT.cpp [hunk 8]
```diff
@@ -16,7 +16,8 @@
 // ./SmallPT 500 && xv image.ppm
 
 // A typical invocation would be:
-// ../../../../../obj/Debug+Asserts/bin/clang++ -O3 -Xclang -add-plugin -Xclang clad \
```
Comment (parth-07) [inline demos/ComputerGraphics/smallpt/SmallPT.cpp:19]: Can you please revert back the changes to the demos?

File: lib/Differentiator/ReverseModeForwPassVisitor.cpp [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+  auto paramTypes = ComputeParamTypes(args);
+  auto returnType = ComputeReturnType();
+  const auto* sourceFnType = dyn_cast<FunctionProtoType>(m_Function->getType());
+  auto fnType = m_Context.getFunctionType(returnType, paramTypes,
+                                          sourceFnType->getExtProtoInfo());
+
+  llvm::SaveAndRestore<DeclContext*> saveContext(m_Sema.CurContext);
+  llvm::SaveAndRestore<Scope*> saveScope(m_CurScope);
+  m_Sema.CurContext = const_cast<DeclContext*>(m_Function->getDeclContext());
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeForwPassVisitor.cpp:43]:
```suggestion
  m_Sema.CurContext = const_cast<DeclContext*>(m_Function->getDeclContext()); NOLINT(cppcoreguidelines-pro-type-const-cast)
```

**Review by vgvassilev — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**